pkg/virt-operator/resource/apply/apps.go

```diff
@@ -17,121 +16,136 @@ import (
-	}
-
 	if !exists {
-		r.expectations.Deployment.RaiseExpectations(kvkey, 1, 0)
-		_, err = apps.Deployments(r.kv.Namespace).Create(context.Background(), deployment, metav1.CreateOptions{})
+		r.expectations.Deployment.RaiseExpectations(r.kvKey, 1, 0)
+		deployment, err := apps.Deployments(kv.Namespace).Create(context.Background(), deployment, metav1.CreateOptions{})
 		if err != nil {
-			r.expectations.Deployment.LowerExpectations(kvkey, 1, 0)
+			r.expectations.Deployment.LowerExpectations(r.kvKey, 1, 0)
 			return fmt.Errorf("unable to create deployment %+v: %v", deployment, err)
 		}
-	} else if !objectMatchesVersion(&cachedDeployment.ObjectMeta, imageTag, imageRegistry, id, r.kv.GetGeneration()) {
-		// Patch if old version
-		var ops []string
 
-		// Add Labels and Annotations Patches
-		labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&deployment.ObjectMeta)
-		if err != nil {
-			return err
-		}
-		ops = append(ops, labelAnnotationPatch...)
+		resourcemerge.SetDeploymentGeneration(&kv.Status.Generations, deployment)
 
-		// Add Spec Patch
-		newSpec, err := json.Marshal(deployment.Spec)
-		if err != nil {
-			return err
-		}
-		ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec", "value": %s }`, string(newSpec)))
+		return nil
+	}
 
-		_, err = apps.Deployments(r.kv.Namespace).Patch(context.Background(), deployment.Name, types.JSONPatchType, generatePatchBytes(ops), metav1.PatchOptions{})
-		if err != nil {
-			return fmt.Errorf("unable to patch deployment %+v: %v", deployment, err)
-		}
-		log.Log.V(2).Infof("deployment %v updated", deployment.GetName())
+	cachedDeployment := obj.(*appsv1.Deployment)
+	modified := resourcemerge.BoolPtr(false)
+	existingCopy := cachedDeployment.DeepCopy()
+	expectedGeneration := resourcemerge.ExpectedDeploymentGeneration(deployment, kv.Status.Generations)
 
-	} else {
+	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, deployment.ObjectMeta)
+	// there was no change to metadata, the generation matched
+	if !*modified && existingCopy.ObjectMeta.Generation == expectedGeneration {
 		log.Log.V(4).Infof("deployment %v is up-to-date", deployment.GetName())
+		return nil
+	}
+
+	// Patch if old version
+	var ops []string
+
+	// Add Labels and Annotations Patches
+	labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&deployment.ObjectMeta)
+	if err != nil {
+		return err
 	}
+	ops = append(ops, labelAnnotationPatch...)
+
+	// Add Spec Patch
+	newSpec, err := json.Marshal(deployment.Spec)
+	if err != nil {
+		return err
+	}
+	ops = append(ops, fmt.Sprintf(replaceSpecPatchTemplate, string(newSpec)))
+
+	deployment, err = apps.Deployments(kv.Namespace).Patch(context.Background(), deployment.Name, types.JSONPatchType, generatePatchBytes(ops), metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to update deployment %+v: %v", deployment, err)
+	}
+
+	resourcemerge.SetDeploymentGeneration(&kv.Status.Generations, deployment)
```

Comment:
I've been pondering this for awhile. There's a (possibly benign) race between when we set this new generation on the kubevirt object and when we actually see the new generation on the kubevirt object in our cache.

So, for instance, if you did the following sequence, you'd get an interesting result.

pseudo code 

```
resourcemerge.SetDeploymentGeneration(&kv.Status.Generations, deployment)
client.Update(kv)
cachedKV := getKVFromCache(kvInformer, kv.Name, kv.Namespace)

if !reflect.DeepEquals(cachedKV, kv) {
     // cache doesn't reflect the latest generation changes.
}
```

I'm not sure what to think of this race condition though. Functionally, it might not cause any problems except that we may Patch the same object a few times with the same change until all the informers catch up. This is the same race that the "expectations" ref counting logic exists to solve.

If applying the same patch to an already patched object never results in the generation ID changing somehow, then I *think* the issue i'm describing here woudl just be an optimization that probably isn't worth our time. If however having the cache not synced with reality can result in generation id of a resource spinning, we'd be in trouble. 

I don't know if i did a good job explaining my thoughts here or not.  I'm about 80% sure it's not an issue. Have you thought about this at all?

Reply:
i don't think my concern here occurs in a way that causes an issue. worst case, we patch and already patched object. When that occurs, the test operation in the patch json will cause the operation to fail until our cache catches up.

Reply:
sorry, I was thinking through this more as well and then didn't submit my comment. 

I agree, the KV object is requested from cache once right before the reconcile loop and then updated at the end of it. Worst case should be that the object is patched, kubernetes sees that nothing is changed and no ops, then returns the same object with the same generation as it was. 

I'm not sure what you are referring to by the `test operation`, we don't add a test operation patch to the JSON patch for the four resources included in this PR. Is that something that should be added?

Reply:
ah! I think it would be wise to add a test op to the patch... Actually, the test operation can be to verify the generation of the cached object matches the generation in our kv.Status.Generations list. 

so something like

``` fmt.Sprintf(`{ "op": "test", "path": "/metadata/generation", "value": %s }`, expectedGenerationID)```

what do you think?

Reply:
wont that then fail when we actually want to update? User updates resource generation ID is increased but the expectedGeneration on the KV resource is not, we want to update the resource but the generation id in cache would then not match the expected and error? Should we instead add the expected generation of the cached object to the test patch?